src/test/java/org/cactoos/map/MapEnvelopeTest.java

```diff
@@ -203,4 +207,103 @@ public void mapContainsValueFalse() {
+        final MapOf<String, String> map =
+            new MapOf<String, String>(new MapEntry<>("key", "value"));
+        MatcherAssert.assertThat(
+            "Map doesn't equal to itself",
+            map,
+            new IsEqual<>(map)
+        );
+    }
+
+    @Test
+    public void mapNotEqualsToAnotherClass() {
+        final MapOf<String, String> map =
+            new MapOf<String, String>(new MapEntry<>("key1", "value1"));
+        MatcherAssert.assertThat(
+            "Map equals to an instance of another type",
+            map,
+            new IsNot<>(
+                new IsEqual<>("Totally different type")
+            )
+        );
+    }
+
+    @Test
+    public void mapEqualsToMapWithSameEntries() {
+        final MapEntry<String, String> entry =
+            new MapEntry<>("key2", "value2");
+        MatcherAssert.assertThat(
+            "Map doesn't equal to another map with same entries",
+            new MapOf<String, String>(entry),
+            new IsEqual<>(new MapOf<String, String>(entry))
+        );
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void equalFailsOnNull() {
+        final MapEntry<String, String> first =
+            new MapEntry<>("key3", "value3");
+        final MapEntry<String, String> second =
+            new MapEntry<>("key4", null);
+        MatcherAssert.assertThat(
+            "Map allows null values, but shouldn't",
+            new MapOf<String, String>(first, second),
+            new IsEqual<>(new MapOf<String, String>(first, second))
+        );
+    }
+
+    @Test
+    public void mapNotEqualsToOtherWithDifferentKeys() {
+        final String value = "value5";
+        MatcherAssert.assertThat(
+            "Map equals to another map with different keys",
+            new MapOf<String, String>(new MapEntry<>("key5", value)),
+            new IsNot<>(
+                new IsEqual<>(
+                    new MapOf<String, String>(
+                        new MapEntry<>("key6", value)
+                    )
+                )
+            )
+        );
+    }
+
+    @Test
+    public void mapNotEqualsToOtherWithDifferentValues() {
+        final String key = "key7";
+        MatcherAssert.assertThat(
+            "Map equals to another map with different values",
+            new MapOf<String, String>(new MapEntry<>(key, "value7")),
+            new IsNot<>(
+                new IsEqual<>(
+                    new MapOf<String, String>(
+                        new MapEntry<>(key, "value8")
+                    )
+                )
+            )
+        );
+    }
+
+    @Test
+    public void hashCodeDependsOnItems() {
+        final MapEntry<String, String> entry =
+            new MapEntry<>("key9", "value9");
+        MatcherAssert.assertThat(
+            "hashCode returns different results for same entries",
+            new MapOf<String, String>(entry).hashCode(),
+            new IsEqual<>(new MapOf<String, String>(entry).hashCode())
+        );
```

Comment:
@proshin-roman Same as above here, just to enforce that we are using two different objects I think we should create another instance of `MapEntry` with the same values.